pt.6
.TH PURITY 6 "18 December 1989"
.SH NAME
purity
\- a general purpose purity test
.SH SYNOPSIS
.B /usr/games/purity [
.I flags
] [
.I testname
]
.SH DESCRIPTION
Purity is an interactive purity test program with a simple,
user interface and datafile format.  For each test, questions
are printed to the your terminal, and you are prompted for
an answer to the current question.  At a prompt, these are
your choices:
.RS
.IP y
Answer "yes" to the question.
.IP n
Answer "no" to the question.
.IP b
Backup one question, if you answered it incorrectly, or
someone is watching you take the test, and you don't (or do)
want to admit a different answer.
.IP r
Redraw the current question.
.IP q
Quit the test, and print the current score.
.IP ?
Print a help screen for the current prompt.
.IP k
.I Kill
a section of the test.  This skips all the questions
of the test until the next subject heading.
.IP a
Toggle answer mode between
.I real
answers and
.I obfuscated
answers.
.I Real
answers print "yes" and "no", while
.I obfuscated
answers are "Maybe" and "maybe".  Obfuscated answers are
preferred if you are shy, and don't want people to be able to
read your answers over your shoulder as you take the test.
.IP d
Toggle dERanGe output.
.IP s
Print your current score on the test you are taking.
.IP l
Toggle score logging.
.RE
.PP
At the end of the test, your score is printed out.  For most
purity tests, lower scores denote more "experience" of the
test material.

.SH FLAGS
These are the command line flags for the test.
.RS
.IP -a
Show 
.I real
answers (i.e. "yes" and "no") instead of
.I obfuscated
ones (i.e. "Maybe" and "maybe") as you answer
the questions.
.IP -d
PrINt THe tESt in DerANgeD pRInT.
.IP -f
Take the test in
.I fast
mode.  Only the questions are printed, and not any other
text blocks, like the introdution, subject headers, and
the conclusion.
.IP -l
Take the test without having your score logged.
.IP -p
Print the test without prompting for answers.  This is
useful for making hard copies of the tests without
having to edit out the prompts by hand.
.IP -r
Decrypt the test using the
.I Rot 13
algorithm.  This is
done as a form of "protection", such that if you read a
rot13 test and it offends you, it's your own fault.
.IP -z
.I zoom
through more prompts in large text blocks.  The
default is to prompt the user for 
.I "more"
when a screenful
of text has been printed without any user input.
.RE
.SH DATAFILE FORMAT
The format of the datafiles is a very simple format,
intended such that new tests can quickly and easily
be converted to run with the test.
.PP
There are four types of text in a purity test datafile.
Each type is contained in a
.I bracket
type of punctuation.  The definitions are as follows:
.PP
the styles of text blocks are:
.IP
{ plain text block }
.IP
[ subject header ]
.IP
( test question )
.IP
and	< conclusion >
.PP
Plain text blocks are printed out character for character.
.PP
Subject headers are preceded by their subject numbers, starting
at 1, and then printed as text blocks.
.PP
Questions are preceded by their numbers, and then prompt the
user to answer the question, keeping track of the user's
current score.
.PP
Conclusions first calculate and print the user's score for the
test, then print out the conclusion as a text block.
.PP
If you wish to include any of the various bracket punctuation in
your text, the backslash ("\\") character will 
.I escape
the next character.

To print a question with parentheses, you would use the following format:
.PP
(have you ever written a purity test \\(like this one\\)?)
.PP
the output would be this:
.PP
   1.  have you ever written a purity test (like this one)?
.PP
and then it would have asked the user for her/his answer.
.PP
For a generic datafile, use the "sample" datafile for the test.
.SH FILES
.nf
/var/games/purity.scores	the score logfile
/usr/share/games/purity/*		test data files
.fi
.SH AUTHOR
Eric Lechner, lechner@ucscb.ucsc.edu


debian/changelog
purity (1-18) unstable; urgency=low

  * QA upload (setting Maintainer to QA group)
  * change to format "3.0 (quilt)"
  * drop upstream changes easily done in debian/rules instead
  * modernize debian/rules
  - use dpkg-buildflags (Closes: 437831)
  - implement build-arch, build-indep
  - debhelper compatibility 7
  * avoid some compiler warnings
  * don't change binary's permissions in postinst but in package
  * Standards-Version to 3.9.2
  * fix some minor formating errors and misspellings lintian warns about

 -- Bernhard R. Link <brlink@debian.org>  Wed, 05 Oct 2011 18:10:50 +0200

purity (1-17) unstable; urgency=low

  * Changed fix for overflow to match that of security releases 1-9.1
    and 1-14.1

 -- Duncan Findlay <duncf@debian.org>  Thu, 12 Sep 2002 19:46:47 -0400

purity (1-16) unstable; urgency=low

  * Prevents possible buffer overflow (Closes: #154290)

 -- Duncan Findlay <duncf@debian.org>  Sun,  8 Sep 2002 22:41:15 -0400

purity (1-15) unstable; urgency=low

  * Change of e-mail address

 -- Duncan Findlay <duncf@debian.org>  Tue, 14 May 2002 19:02:47 -0400

purity (1-14) unstable; urgency=low

  * Change of e-mail address

 -- Duncan Findlay <duncf@rogers.com>  Sat, 24 Nov 2001 16:11:01 -0500

purity (1-13) unstable; urgency=low

  * fixed segmentation fault when saving high scores (closes: #114340) -
    thanks to Gergely Nagy for the patch 

 -- Duncan Findlay <duncf@home.com>  Wed,  3 Oct 2001 22:12:24 -0400

purity (1-12) unstable; urgency=low

  * new maintainer (closes: #113109)
  * fixed stty settings munging (closes: #112827, #112830) - thanks to
    Alan Shutko for the patch
  * added build depends (closes: #107311, #111794)
  * moved /var/lib/games to /var/games (closes: #107643)
  * (slightly) improved description (closes: #112806)

 -- Duncan Findlay <duncf@home.com>  Thu, 20 Sep 2001 22:54:11 -0400

purity (1-11) unstable; urgency=low

  * changed package from extra to optional
  * make directory /var/lib/games if it doesn't exist. (closes #107015)

 -- Aaron Howell <aaron@kitten.net.au>  Mon, 30 Jul 2001 07:55:40 +1000

purity (1-10) unstable; urgency=low

  * Fixed Readme file (closes #48608).
  * Applied patch from James Morrison to make purity build on SYSV
    platforms using termios.

 -- Aaron Howell <aaron@kitten.net.au>  Sun, 29 Jul 2001 17:15:53 +1000

purity (1-9) unstable; urgency=low

  * converted to debhelper
  * changed sunsite.unc.edu to metalab.unc.edu (fixes #33612)
  * moved tests under /usr/share/games/purity (fixes #44912)

 -- Aaron Howell <aaron@kitten.net.au>  Tue, 21 Sep 1999 10:39:15 +1000

purity (1-8) unstable; urgency=low

  * fixed line length of extended description to conform to policy,
    fixes #30086
  * changed maintainer address to aaron@kitten.net.au

 -- Aaron Howell <aaron@kitten.net.au>  Sat, 9 Jan 1999 15:25:59 +1000

purity (1-7) unstable; urgency=low

  * change maintainer address back to cnl, too many problems with mail to bayside.

 -- Aaron Howell <aaron@sunrise.cnl.com.au>  Sat, 8 Aug 1998 18:44:27 +1000

purity (1-6) unstable; urgency=low

  * fix source and postinst so scores are logged to /var/lib/games/purity.scores (#18586)

 -- Aaron Howell <aaron@sunrise.bayside.net.au>  Wed, 13 May 1998 10:27:03 +1000

purity (1-5) unstable; urgency=low

  * s/cnl.com.au/bayside.net.au/g

  * rebuild to fix md5sums-contains-nonexisting-file error from lintian fixes #19394

 -- Aaron Howell <aaron@sunrise.bayside.net.au>  Wed, 11 Mar 1998 16:18:27 +1000

purity (1-4) unstable; urgency=low

  * fix menu entry so purity gets started with a test that is actually included with the package. fixes #16478

 -- Aaron Howell <aaron@sunrise.cnl.com.au>  Sat,  3 Jan 1998 02:48:31 +1000

purity (1-3) unstable; urgency=low

  * removed offensive purity tests in accordance with recommendations on debian-devel, see README.debian for more information

 -- Aaron Howell <aaron@sunrise.cnl.com.au>  Sun, 30 Nov 1997 15:19:59 +1000

purity (1-2) unstable; urgency=low

  * reupload, hopefully this time my public key will be accepted

 -- Aaron Howell <aaron@sunrise.cnl.com.au>  Sat, 18 Oct 1997 23:30:59 +1000

purity (1-1) unstable; urgency=low

  * Changed termios.h to asm/termios.h so  purity will build ok.
  * Initial Release.

 -- Aaron Howell <aaron@sunrise.cnl.com.au>  Mon, 29 Sep 1997 10:30:17 +1000


pt.c
/*
	pt.c   version 3.9   December 19, 1989

	Eric Lechner  (lechner@ucscb.ucsc.edu)

	a general purpose N-question (purity) test.
	data file format given in pt.h
	tailor definitions to your needs in pt.h
*/

#define VERSION "3.9" 			/* current test version # */
#ifdef LINUX
# define SYSV
# include <unistd.h>
# include <stdlib.h>
#endif

#include "pt.h"
#include <ctype.h>
#include <stdio.h>
#ifdef SYSV
#include <sys/types.h>
#include <sys/stat.h>
#include <termios.h>  /* SYSV now uses termios instead of termio POSIX compliant */
#include <string.h>
#else
#include <sgtty.h>
#include <strings.h>
#endif /*SYSV*/
#include <sys/ioctl.h>
#include <sys/file.h>
#include <signal.h>
#include <pwd.h>

#define SKIP 255
#define TRUE 1
#define FALSE 0
#define DO_ECHO TRUE
#define NO_ECHO FALSE
#define OK 10
#define FILE_ERR -1
#define QUIT -2
#define EXIT -3
#define BACKUP 5
#define LOOP -10

struct quest {
	int num,		/* # of question		*/
	    answer;		/* answer to current question	*/
	long fpos;		/* position in file for fseek	*/
	struct quest *last;	/* pointer to last question	*/
};

static int	question(void),
	get_response(void),
	getresp(void),
	subject(void),
	print(int, int),
	more(void),
	redraw(void);

static void	explain(void),
	quit(int),
#ifdef SIGTSTP
	ctrlz(int),
	ctrlzret(int),
#endif
	set_mode(void),
	clr_mode(void),
	freemem(void),
	printscore(void),
	sizewindow(void);

#ifndef PW
struct passwd *getpw();
#endif

static FILE *fp;

static struct quest *theq, *nextq, *the_s, *next_s;
static int num_ch, num_ln, kill_sub, no_ans, zoom, obfus, fast, derange, rot13;

static short tty_flags;
#ifdef SYSV
static unsigned char baz;
#endif

static unsigned short rows, cols;

static char *testtype;
#ifdef LOGFILE
static int logme;
#endif

int main(int argc, char *argv[])
{
	int err, ch;	/* var to store error reports */
	char path[256], *ch1;

	(void) signal(SIGINT,quit);
#ifdef SIGTSTP
	(void) signal(SIGTSTP,ctrlz);
	(void) signal(SIGCONT,ctrlzret);
#endif

#ifdef GID
	(void) setgid(GID);
#endif
	srand (getpid());	/* seed r.n.g. for derange mode */
				/* it's only derange, why not use the pid? */

	err = 0;
	*path = '\0';
	testtype = NULL;
	ch = FALSE;

	while (ch == FALSE) {
		if (++err >= argc) {
			(void) sprintf(path,"%s/%s",LIBDIR,"intro");
			ch = TRUE;
		} else {
			switch (*argv[err]) {
				case '-' :
					break;
				default :
					(void) snprintf(path, sizeof(path),"%s/%s",
							LIBDIR, argv[err]);
					ch = TRUE;
					break;
			}
		}
	}

	fp = fopen(path,"r");
	if (fp == NULL) {
		(void) snprintf(path, sizeof (path),"%s",argv[err]);
		fp = fopen(path,"r");
	}

	if (fp == NULL) {
#ifndef SYSV
		if ((ch1 = rindex(argv[0],'/'))) ch1++;
#else
		if ((ch1 = strrchr(argv[0],'/'))) ch1++;
#endif
		else ch1 = argv[0];
#ifndef SYSV
		if ((testtype = rindex(path,'/'))) testtype++;
#else
		if ((testtype = strrchr(path,'/'))) testtype++;
#endif
		else testtype = path;
		
		fprintf(stderr,"%s: error opening datafile '%s'.\n",
			ch1,testtype);
		exit(-1);
	}
	rewind (fp);

	if (testtype == NULL) {
#ifndef SYSV
		if ((testtype = rindex(path,'/'))) testtype++;
#else
		if ((testtype = strrchr(path,'/'))) testtype++;
#endif
		else testtype = path;
	}

#ifdef LOGFILE

	if (!strcmp(testtype,"intro") ||
			!strcmp(testtype,"scores") ||
			!strcmp(testtype,"format") ||
			!strcmp(testtype,"help") ||
			!strcmp(testtype,"list"))
		logme = FALSE;
	else logme = TRUE;
#endif

	theq = (struct quest *) malloc (sizeof(struct quest));
	the_s = (struct quest *) malloc (sizeof(struct quest));
	if ((theq == NULL) || (the_s == NULL)) {
		fprintf(stderr,"Memory allocation error.\n");
		(void) fclose(fp);
		exit(-1);
	}

#ifdef MASK
	(void) strncpy (argv[0],MASK,strlen(argv[0]));
#endif

	theq->num = 1;		/* setup initial question pointer info */
	theq->last = NULL;
	theq->answer = FALSE;
	the_s->num = 1;		/* setup initial subject pointer info */
	the_s->last = NULL;
	no_ans = zoom = fast = derange = rot13 = FALSE;	/* set default flags */
	obfus = TRUE;


	if (argc > 1) {		/* read command line options, set flags */
		for (err=1; err<argc; err++) {
			if ((*argv[err] == '-') || (*argv[err] == '+')) {
				while (*++argv[err]) 
				switch (*argv[err]) {
					case 'p' :
						no_ans = TRUE;
						break;
					case 'z' :
						zoom = TRUE;
						break;
#ifdef LOGFILE
					case 'l' :
						logme = FALSE;
						break;
#endif
					case 'a' :
						obfus = FALSE;
						break;
					case 'f' :
						fast = TRUE;
						break;
					case 'd' :
						derange = TRUE;
						break;
					case 'r' :
						rot13 = TRUE;
						break;
					default : break;
				}
			}
		}
	}
		
	set_mode();		/* set no-echo and cbreak modes	*/
	sizewindow();		/* get window size for "more" prompts */
	num_ch = num_ln = 0;
	kill_sub = FALSE;

#ifdef LOGFILE
	if (strcmp(testtype,"scores"))
#endif
		err = OK;
#ifdef LOGFILE
	else {
		print(DO_ECHO,EOF);
		err = EXIT;
	}
#endif

	while (err == OK) {	/* do all the test stuff... */
		ch = fgetc(fp);
		switch (ch) {
			case EOF :
				err = EXIT;
				break;
			case TEXT :
				if (fast) err = print(NO_ECHO,ENDTEXT);
				else {
					err = print(DO_ECHO,ENDTEXT);
					(void) fputc('\n',stdout);
				}
				break;
			case SUBJECT :
				err = subject();
				break;
			case QUESTION :
				err = question();
				break;
			case PRINTSCORE :
				printscore();		/* print the score */
				if (!fast) err = print(DO_ECHO,ENDSCORE);
				else err = print(NO_ECHO,ENDSCORE);
				break;
			default :
				/*fputc(ch,stdout);*/
				break;
		}
	}
	(void) fclose (fp);
	(void) fflush(stdout);
	clr_mode();
	switch (err) {
		case FILE_ERR :
			fprintf(stderr,"\nFile access error.\n");
			freemem();
			exit (-1);
			break;
		case QUIT :
			printscore();
			freemem();
			fprintf (stdout,"\nseeya later, alligator.\n");
			exit(0);
			break;
		case EXIT :
			fprintf (stdout,"bye.\n");
			freemem();
			exit(0);
			break;
		default :
			break;
	}
	return 0;
}

int question()
{
	int err;

	theq->fpos = ftell(fp) - 1;

	if (kill_sub == FALSE) {
		fprintf(stdout,"%3d. ",theq->num);
		err = print(DO_ECHO,ENDQUEST);
		(void) fputc('\n',stdout);
	} else err = print(NO_ECHO,ENDQUEST);

	if (err == OK) {
		if (kill_sub || no_ans) err = SKIP;
		else err = get_response();

		switch (err) {
			case EXIT :
			case QUIT :
				break;
			case BACKUP :
				if (theq->last != NULL) {
					nextq = theq;
					theq = nextq->last;
					free((char *) nextq);
					fprintf(stdout,"Your old answer was: ");
					if (theq->answer == TRUE)
						fprintf(stdout,"yes.\n\n");
					else if (theq->answer == FALSE)
						fprintf(stdout,"no.\n\n");
					else fprintf(stdout,"skipped.\n\n");
					if (the_s->last != NULL) {
						next_s = the_s->last;
						if (next_s->fpos > theq->fpos) {
							free((char *) the_s);
							the_s = next_s;
						}
					}
				} else fprintf(stderr,"Silly, you can't back up past the beginning of the test...\n\n");
				err = fseek(fp,theq->fpos,0);
				if (err == EOF) err = FILE_ERR;
				else err = OK;
				break;
			default :
				theq->answer = err;
				err = OK;
				nextq = (struct quest *) malloc (sizeof(struct quest));
				if (nextq == NULL) {
					fprintf(stderr,"Memory allocation error.\n");
			 		err = EXIT;
				} else {
					nextq->num = theq->num + 1;
					nextq->last = theq;
					nextq->answer = FALSE;
					theq = nextq;
				}
				break;
		}
	}
	if (no_ans == FALSE) num_ch = num_ln = 0;
	return(err);
}

int get_response()
{
	int ch, done = FALSE;

	while (done == FALSE) {
#ifndef LOGFILE
		fprintf(stdout,"your answer? [ynbakdsrq?] : ");
#else
		fprintf(stdout,"your answer? [ynbakdlsrq?] : ");
#endif

		ch = getresp();

		switch (ch) {
			case 'y' :
			case 'Y' :
			case ' ' :
			case '\n' :
				if (obfus)
					fprintf(stdout,"%s\n\n",YES);
				else fprintf(stdout,"yes\n\n");
				ch = TRUE;
				done = TRUE;
				break;
			case 'n' :
			case 'N' :
				if (obfus)
					fprintf(stdout,"%s\n\n",NO);
				else fprintf(stdout,"no\n\n");
				ch = FALSE;
				done = TRUE;
				break;
			case 'b' :
			case 'B' :
			case '-' :
			case 'c' :
			case 'C' :
				fprintf(stdout,"backup\n\n");
				ch = BACKUP;
				done = TRUE;
				break;
			case 'x' :
			case 'X' :
			case 'q' :
			case 'Q' :
				fprintf(stdout,"quit\n");
				ch = QUIT;
				done = TRUE;
				break;
			case 's' :
			case 'S' :
				fprintf(stdout,"print score\n");
				printscore();
				ch = 's';
			case 'r' :
			case 'R' :
				if (ch != 's') fprintf(stdout,"redraw\n");
				(void) fputc('\n',stdout);
				ch = redraw();
				if (ch == FILE_ERR) {
					done = TRUE;
					ch = EXIT;
				}
				break;
			case 'h' :
			case 'H' :
			case '?' :
				fprintf(stdout,"help\n\n");
				explain();
				break;
			case 'k' :
			case 'K' :
				fprintf(stdout,"kill\n\n");
				kill_sub = TRUE;
				ch = SKIP;
				done = TRUE;
				break;
#ifdef LOGFILE
			case 'l' :
			case 'L' :
				fprintf (stdout,"logscore is now ");
				if (logme) {
					logme = FALSE;
					fprintf (stdout,"off\n");
				} else {
					logme = TRUE;
					fprintf(stdout,"on\n");
				}
				break;
#endif
			case 'd' :
			case 'D' :
				fprintf (stdout,"derange is now ");
				if (derange) {
					derange = FALSE;
					fprintf (stdout,"off\n");
				} else {
					derange = TRUE;
					fprintf(stdout,"on\n");
				}
				break;
			case 'a' :
			case 'A' :
				fprintf (stdout,"answers are now ");
				if (obfus) {
					obfus = FALSE;
					fprintf(stdout,"real\n");
				} else {
					obfus = TRUE;
					fprintf (stdout,"obfuscated\n");
				}
				break;
			default :
				fprintf(stdout,"huh?\n");
				break;
		}
	}
	return (ch);
}

void explain()
{
	fprintf(stdout,"the purity test: version %s  by Eric Lechner\n\n",
								VERSION);
	fprintf(stdout,"y - answer 'yes' to current question.\n");
	fprintf(stdout,"n - answer 'no' to current question.\n");
	fprintf(stdout,"b - backup one question.\n");
	fprintf(stdout,"a - answer style (%s/%s vs. yes/no)\n",YES,NO);
	fprintf(stdout,"k - kills all questions until the next subject.\n");
	fprintf(stdout,"d - toggles between normal and dERanGeD text.\n");
	fprintf(stdout,"s - prints your current score for the test.\n");
#ifdef LOGFILE
	fprintf(stdout,"l - toggles whether or not your score is logged.\n");
#endif
	fprintf(stdout,"r - redraw current question.\n");
	fprintf(stdout,"q - quit the program.\n");
	fprintf(stdout,"? - print this help screen.\n\n");
}

int getresp()
{
	int ch;

	ch = fgetc(stdin);

	return (ch);
}

int subject()
{
	int err;

	the_s->fpos = ftell(fp) - 1;
	kill_sub = FALSE;

	if (fast) err = print (NO_ECHO,ENDSUB);
	else {
		fprintf(stdout,"------------------------------------------------------------------------------\n");
		fprintf(stdout,"%d. ",the_s->num);
		err = print(DO_ECHO,ENDSUB);
		(void) fputc('\n',stdout);
	}
	if (err == OK) {
		next_s = (struct quest *) malloc (sizeof(struct quest));
		if (next_s == NULL) {
			fprintf(stderr,"Memory allocation error.\n");
			err = EXIT;
		} else {
			next_s->last = the_s;
			next_s->num = the_s->num + 1;
			the_s = next_s;
		}
	}
	return (err);
}

int print(int echo, int esc)
	/* if echo then print the text, and give "more" prompts */
	/* the escape character to end this text block */
{
	int ch;

	while (TRUE) {
		ch = fgetc(fp);
		if (ch == EOF) return (FILE_ERR);
		if (ch == '\\') ch = fgetc(fp);
		else if (ch == esc) {	/* check next char for escapeness */
			return(OK);
		}
		if (echo) {
			if (derange)
				if ((ch >= 'a') && (ch <= 'z'))
					if (rand() % 2) ch += 'A' - 'a';
			if (rot13) {
				if ((ch <= 'z') && (ch >= 'a')) {
					if ((ch += 13) > 'z') ch -= 26;
				} else if ((ch <= 'Z') && (ch >= 'A')) {
					if ((ch += 13) > 'Z') ch -= 26;
				}
			}
			(void) fputc((char)ch,stdout);
			if (zoom == FALSE) {
				if (ch == '\n') {
					num_ch = 0;
					num_ln++;
				}
				if (num_ch >= cols) {	/* autowrap */
					num_ln++;
					(void) fputc('\n',stdout);
				}
				if (num_ln >= rows - 2) echo = more();
				if (echo == QUIT) return (QUIT);
			}
		}
	}
}

int more()
{
	int resp;
	long ret = LOOP;

	num_ch = num_ln = 0;
	fprintf(stdout,"--- more ---");
	while (ret == LOOP) {
		resp = getresp();
		switch (resp) {
			case ' ' :
			case 'y' :
			case 'Y' :
			case 'm' :
			case 'M' :
			case '\n' :
				ret = DO_ECHO;
				break;
			case 'n' :
			case 'N' :
			case 'k' :
			case 'K' :
				ret = NO_ECHO;
				break;
			case 'q' :
			case 'Q' :
			case 'x' :
			case 'X' :
				ret = QUIT;
				break;
			default :
				(void) fputc(007,stdout);
				break;
		}
	}
	fprintf(stdout,"\b\b\b\b\b\b\b\b\b\b\b\b            \b\b\b\b\b\b\b\b\b\b\b\b");
	return (ret);
}

void quit (int s __attribute__((unused)))
{
	(void) fflush(stdout);
	(void) signal (SIGINT, SIG_DFL);
	clr_mode();
	printscore();
	freemem();
	fprintf(stderr,"\nFine. Be that way. Don't use the 'q' option.\n");
	exit (0);
}

#ifdef SIGTSTP
void ctrlz (int s __attribute__((unused)))
{
	(void) fflush(stdout);
	fprintf(stdout,"\nyou'll be back, right?\n");
	clr_mode();
	(void) signal (SIGTSTP, SIG_DFL);
	(void) kill (getpid(), SIGTSTP);
}

void ctrlzret (int s __attribute__((unused)))
{
	fprintf(stdout,"nice to see you came back...\n");
	(void) fflush(stdout);
	set_mode();
}
#endif

void set_mode()
{
#ifndef SYSV
	struct sgttyb p;
#else
	struct termios p;
#endif

#ifndef SYSV
	(void) ioctl(0,TIOCGETP,&p);
	tty_flags = p.sg_flags;
	p.sg_flags |= CBREAK;
	p.sg_flags &= ~ECHO;
	(void) ioctl(0,TIOCSETP,&p);
#else
	tcgetattr(0,&p);
	tty_flags = p.c_lflag;
	baz = p.c_cc[VMIN];
	p.c_lflag &= ~ICANON;
	p.c_lflag &= ~ECHO;
	p.c_cc[VMIN] = 1;
	tcsetattr(0,TCSANOW,&p);
#endif
}

void clr_mode()
{
#ifndef SYSV
	struct sgttyb p;
#else
	struct termios p;
#endif

#ifndef SYSV
	(void) ioctl(0,TIOCGETP,&p);
	p.sg_flags = tty_flags;
	(void) ioctl(0,TIOCSETP,&p);
#else
	tcgetattr(0,&p);
	p.c_lflag = tty_flags;
	p.c_cc[VMIN] = baz;
	tcsetattr(0,TCSANOW,&p);
#endif
}

void printscore()
{
	int score = 0, num_q = 0;
	if (no_ans) return;

	nextq = theq->last;
	while (nextq != NULL) {
		if (nextq->answer == FALSE) score++;
		if (nextq->answer != SKIP) num_q++;
		nextq = nextq->last;
	}
	if (num_q <= 0) {
		fprintf(stdout,"\nyou can't get a score without answering any questions, silly.\n\n");
		num_ln += 2;
	} else {
		fprintf(stdout, "\nyou answered %d 'no' answer%s out of %d question%s,\n",score,(score -1) ? "s" : "",num_q,(num_q-1) ? "s" : "");
		fprintf(stdout,"which makes your purity score %.2f%%.\n\n",
			((float) 100 * score / num_q));
		num_ln += 3;
	}
}

void freemem()
{
#ifdef LOGFILE
	int score = 0, ans = 0;
	char login[8], scorepath[512];
	struct passwd *pw;
#endif

	while (theq != NULL) {
		nextq = theq->last;
		free ((char *) theq);
		theq = nextq;
#ifdef LOGFILE
		if ((logme) && (no_ans == FALSE)) {
			if (theq != NULL) {
				if (nextq->answer == FALSE) score++;
				if (nextq->answer != SKIP) ans++;
			}
		}
#endif
	}
	while (the_s != NULL) {
		next_s = the_s->last;
		free ((char *) the_s);
		the_s = next_s;
	}

#ifdef LOGFILE
	if ((no_ans == FALSE) && logme && (ans > 0)) {
# ifdef SCOREFILE
		strcpy(scorepath, SCOREFILE);
# else
		sprintf(scorepath,"%s/%s",LIBDIR,"scores");
# endif
		fp = fopen (scorepath,"a+");
		if (fp != NULL) {
#ifdef GETLOGIN
			char *tmp = getlogin();
			(void) strcpy(login,(tmp) ? tmp : "");
			if (*login == '\0') {
#endif
#ifdef PW
				pw = getpwuid(getuid());
#else
				pw = getpw(getuid());
#endif
				(void) strcpy(login,pw->pw_name);
#ifdef GETLOGIN
			}
#endif
			fprintf (fp,"%s:%s: %d / %d = %.2f%%\n",login,testtype,
				score, ans, (float) ((100 * score) / ans));
			(void) fclose (fp);
		}
	}
#endif
}

int redraw()
{
	int err;

	err = fseek(fp,theq->fpos+1,0);
	if (err == -1) err = FILE_ERR;
	else {
		fprintf(stdout,"%3d. ",theq->num);
		err = print(DO_ECHO,ENDQUEST);
		fprintf(stdout,"\n");
	}

	return(err);
}

void sizewindow()
{
	struct winsize wsize;

	(void) ioctl(0,TIOCGWINSZ,&wsize);

	cols = wsize.ws_col;
	rows = wsize.ws_row;

	if (cols == 0) cols = 80;	/* if no row/col values, assign	*/
	if (rows == 0) rows = 24;	/* reasonable default ones.	*/
}

#ifndef PW
struct passwd *getpw(uid)
int uid;
{
	FILE *pwfp;	/* the password file pointer */
	struct passwd pw;
	char line[1024], *tmp, *tmp2;
	int count, ch;

	pwfp = fopen("/etc/passwd","r");
	if (pwfp == NULL) {
		return ((struct passwd *) NULL);
	} else {
		ch = fgetc(pwfp);
		count = 0;
		while (ch != EOF) {
			if (ch != '\n') {
				line[count++] = ch;
			} else {
				line[count] = '\0';
				count = 0;
				pw.pw_name = line;
#ifndef SYSV
				tmp = index(pw.pw_name,':');
				*tmp = '\0';
				pw.pw_passwd = ++tmp;
				tmp = index(pw.pw_passwd,':');
				*tmp = '\0';
				tmp2 = ++tmp;
				tmp = index(tmp2,':');
				*tmp = '\0';
				pw.pw_uid = atoi(tmp2);
				tmp2 = ++tmp;
				tmp = index(tmp2,':');
				*tmp = '\0';
				pw.pw_gid = atoi(tmp2);
				pw.pw_comment = ++tmp;
				tmp = index(pw.pw_comment,':');
				*tmp = '\0';
				pw.pw_dir = ++tmp;
				tmp = index(pw.pw_dir,':');
#else
				tmp = strchr(pw.pw_name,':');
				*tmp = '\0';
				pw.pw_passwd = ++tmp;
				tmp = strchr(pw.pw_passwd,':');
				*tmp = '\0';
				tmp2 = ++tmp;
				tmp = strchr(tmp2,':');
				*tmp = '\0';
				pw.pw_uid = atoi(tmp2);
				tmp2 = ++tmp;
				tmp = strchr(tmp2,':');
				*tmp = '\0';
				pw.pw_gid = atoi(tmp2);
				pw.pw_comment = ++tmp;
				tmp = strchr(pw.pw_comment,':');
				*tmp = '\0';
				pw.pw_dir = ++tmp;
				tmp = strchr(pw.pw_dir,':');
#endif
				*tmp = '\0';
				pw.pw_shell = ++tmp;
				if (pw.pw_uid == uid) {
					fclose(pwfp);
					return (&pw);
				}
			}
			ch = fgetc(pwfp);
		}
		fclose(pwfp);
		return(NULL);
	}
}
#endif
